### Dataset -  CIFAR-10
* 50,000 training images
* 10,000 test images
* 32x32 pixels RGB

importing libraries

In [3]:
import os
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers
from tensorflow.keras.datasets import cifar10

In [4]:
(x_train,y_train),(x_test,y_test) = cifar10.load_data()
print(x_train.shape,x_test.shape,y_train.shape,y_test.shape)

(50000, 32, 32, 3) (10000, 32, 32, 3) (50000, 1) (10000, 1)


* Here, we have 50,000 datasets for training which is having 32 x 32 pixels , RGB -3
* 10,000 as a test Datasets

#### convering the pixel value from 0 to 1 by dividing the value with 255.0

In [5]:
x_train = x_train.astype(float)/255.0
x_test = x_test.astype(float)/255.0

In [10]:
model = keras.Sequential(
    [
        keras.Input(shape=(32,32,3)),
        layers.Conv2D(32,3,padding='valid',activation='relu'),
        layers.MaxPooling2D(pool_size=(2,2)),
        layers.Conv2D(64,3,activation='relu'),
        layers.MaxPooling2D(),
        layers.Conv2D(128,3,activation='relu'),
        layers.Flatten(),
        layers.Dense(64,activation='relu'),
        layers.Dense(10),
    ]
)
model.summary()

Model: "sequential_3"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ conv2d_11 (Conv2D)              │ (None, 30, 30, 32)     │           896 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_8 (MaxPooling2D)  │ (None, 15, 15, 32)     │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_12 (Conv2D)              │ (None, 13, 13, 64)     │        18,496 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_9 (MaxPooling2D)  │ (None, 6, 6, 64)       │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_13 (Conv2D)              │ (None, 4, 4, 128)      │        73,856 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ flatten_3 (Flatten)             │ (None, 2048)           │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_6 (Dense)                 │ (None, 64)             │       131,136 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_7 (Dense)                 │ (None, 10)             │           650 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 225,034 (879.04 KB)

 Trainable params: 225,034 (879.04 KB)

 Non-trainable params: 0 (0.00 B)

In [14]:
model.compile(
    loss=keras.losses.SparseCategoricalCrossentropy(from_logits=True),
    optimizer = keras.optimizers.Adam(learning_rate=3e-4),
    metrics =['accuracy'],
)

In [17]:
model.fit(x_train,y_train,batch_size=64,epochs=20,verbose=1)
model.evaluate(x_test,y_test,batch_size=64,verbose=1)

Epoch 1/20
782/782 ━━━━━━━━━━━━━━━━━━━━ 12s 16ms/step - accuracy: 0.7350 - loss: 0.7727
Epoch 2/20
782/782 ━━━━━━━━━━━━━━━━━━━━ 11s 15ms/step - accuracy: 0.7469 - loss: 0.7406
Epoch 3/20
782/782 ━━━━━━━━━━━━━━━━━━━━ 12s 15ms/step - accuracy: 0.7527 - loss: 0.7172
Epoch 4/20
782/782 ━━━━━━━━━━━━━━━━━━━━ 12s 16ms/step - accuracy: 0.7629 - loss: 0.6928
Epoch 5/20
782/782 ━━━━━━━━━━━━━━━━━━━━ 12s 15ms/step - accuracy: 0.7737 - loss: 0.6606
Epoch 6/20
782/782 ━━━━━━━━━━━━━━━━━━━━ 12s 15ms/step - accuracy: 0.7813 - loss: 0.6430
Epoch 7/20
782/782 ━━━━━━━━━━━━━━━━━━━━ 12s 16ms/step - accuracy: 0.7926 - loss: 0.6133
Epoch 8/20
782/782 ━━━━━━━━━━━━━━━━━━━━ 13s 17ms/step - accuracy: 0.7991 - loss: 0.5864
Epoch 9/20
782/782 ━━━━━━━━━━━━━━━━━━━━ 15s 19ms/step - accuracy: 0.8062 - loss: 0.5658
Epoch 10/20
782/782 ━━━━━━━━━━━━━━━━━━━━ 16s 20ms/step - accuracy: 0.8151 - loss: 0.5455
Epoch 11/20
782/782 ━━━━━━━━━━━━━━━━━━━━ 15s 19ms/step - accuracy: 0.8241 - loss: 0.5133
Epoch 12/20
782/782 ━━━━━━━━━━

[0.9022545218467712, 0.7322999835014343]

* From the above model it seems that we got 88% accuracy in training data , but in test data we got only 73% accuracy.
* so this model is overfitting (low bais, high variance)

#### Now, we are defining a function to create functional API model which return a model.

In [22]:
def my_model():
    inputs = keras.Input(shape=(32,32,3))
    x = layers.Conv2D(32,3)(inputs)
    x = layers.BatchNormalization()(x)
    x = keras.activations.relu(x)
    x = layers.MaxPooling2D()(x)
    x = layers.Conv2D(64,5,padding='same')(x)
    x = layers.BatchNormalization()(x)
    x = keras.activations.relu(x)
    x = layers.Conv2D(123,3)(x)
    x = keras.activations.relu(x)
    x = layers.Flatten()(x)
    x = layers.Dense(64,activation='relu')(x)
    outputs = layers.Dense(10)(x)
    model = keras.Model(inputs = inputs,outputs=outputs)
    return model

In [23]:
model_1 = my_model()

In [24]:
model_1.compile(
    loss=keras.losses.SparseCategoricalCrossentropy(from_logits=True),
    optimizer = keras.optimizers.Adam(learning_rate=3e-4),
    metrics =['accuracy'],
)

In [25]:
model_1.fit(x_train,y_train,epochs=20,batch_size=64,verbose=0)
model_1.evaluate(x_test,y_test,batch_size=64,verbose=0)

[1.7797240018844604, 0.6858999729156494]

In [26]:
model_1.fit(x_train,y_train,batch_size=64,epochs=20,verbose=1)
model_1.evaluate(x_test,y_test,batch_size=64,verbose=1)

Epoch 1/20
782/782 ━━━━━━━━━━━━━━━━━━━━ 76s 97ms/step - accuracy: 0.9870 - loss: 0.0411
Epoch 2/20
782/782 ━━━━━━━━━━━━━━━━━━━━ 60s 77ms/step - accuracy: 0.9879 - loss: 0.0364
Epoch 3/20
782/782 ━━━━━━━━━━━━━━━━━━━━ 57s 73ms/step - accuracy: 0.9876 - loss: 0.0377
Epoch 4/20
782/782 ━━━━━━━━━━━━━━━━━━━━ 58s 74ms/step - accuracy: 0.9898 - loss: 0.0303
Epoch 5/20
782/782 ━━━━━━━━━━━━━━━━━━━━ 60s 77ms/step - accuracy: 0.9893 - loss: 0.0341
Epoch 6/20
782/782 ━━━━━━━━━━━━━━━━━━━━ 61s 78ms/step - accuracy: 0.9895 - loss: 0.0309
Epoch 7/20
782/782 ━━━━━━━━━━━━━━━━━━━━ 64s 82ms/step - accuracy: 0.9860 - loss: 0.0417
Epoch 8/20
782/782 ━━━━━━━━━━━━━━━━━━━━ 61s 78ms/step - accuracy: 0.9921 - loss: 0.0256
Epoch 9/20
782/782 ━━━━━━━━━━━━━━━━━━━━ 63s 80ms/step - accuracy: 0.9900 - loss: 0.0292
Epoch 10/20
782/782 ━━━━━━━━━━━━━━━━━━━━ 63s 80ms/step - accuracy: 0.9897 - loss: 0.0307
Epoch 11/20
782/782 ━━━━━━━━━━━━━━━━━━━━ 64s 81ms/step - accuracy: 0.9903 - loss: 0.0305
Epoch 12/20
782/782 ━━━━━━━━━━

[2.289672374725342, 0.7081999778747559]

* the model definetly overfitting because it have accuracy of 99% in training and 71% in testing data which shows ( low bias and high variance)